File: leetcode/0872.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<vector<int>> s(2, vector<int>());
        auto dfs = [&](auto&& dfs, TreeNode* root, int f) {
            if (root == nullptr) return;
            if (root->left == nullptr && root->right == nullptr) {
                s[f].push_back(root->val);
            } else {
                dfs(dfs, root->left, f);
                dfs(dfs, root->right, f);
            }
        };
        dfs(dfs, root1, 0);
        dfs(dfs, root2, 1);
        return s[0] == s[1];
    }
};

File: leetcode/0872.go
package leetcode

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	s := make([][]int, 2)
	for i := range s {
		s[i] = make([]int, 0)
	}
	var dfs func(*TreeNode, int)
	dfs = func(root *TreeNode, f int) {
		if root == nil {
			return
		}
		if root.Left == nil && root.Right == nil {
			s[f] = append(s[f], root.Val)
		} else {
			dfs(root.Left, f)
			dfs(root.Right, f)
		}
	}
	dfs(root1, 0)
	dfs(root2, 1)
	if len(s[0]) != len(s[1]) {
		return false
	}
	for i, v := range s[0] {
		if v != s[1][i] {
			return false
		}
	}
	return true
}
